File: apps/desktop-pdv/src/pdv/provisioning/activation.py
"""Ativação do terminal — o pareamento com a retaguarda.

O lojista digita um **código de ativação** curto, gerado no painel administrativo
e válido por poucos minutos. Em troca, o terminal recebe a identidade com que vai
existir na nuvem: `tenant_id`, `store_id`, `device_id` e o token de
sincronização.

Por que um código curto e não o token direto
--------------------------------------------

O token de sincronização é longo e vitalício. Um código de uso único e de vida
curta pode ser ditado por telefone sem que isso comprometa nada de duradouro: se
vazar, ele expira antes de servir para algo, e queimá-lo custa um clique no
painel.

O que **nunca** trafega
-----------------------

O `device_secret` (chave HMAC do ledger) é gerado localmente e não sai da
máquina — ver `secrets.py`. A ativação transporta identidade e autorização, não
a chave que torna a auditoria verificável.

A regra que impede o pior erro de campo
---------------------------------------

**Um terminal com vendas não sincronizadas não pode trocar de tenant.** Aquelas
vendas foram registradas sob o CNPJ antigo; reapontar o terminal antes de
esvaziar a fila faria o faturamento de uma loja desembarcar na outra. É o tipo
de erro que ninguém percebe no dia e que aparece na conciliação fiscal do mês.
Por isso `activate()` recusa a troca com fila pendente, e o operador precisa
sincronizar (ou o suporte precisa decidir conscientemente) antes.
"""

from __future__ import annotations

import logging
import platform
import re
from dataclasses import dataclass
from typing import Protocol

from pdv.data.database import Database
from pdv.data.settings import SettingsStore
from pdv.domain.errors import PdvError
from pdv.provisioning.secrets import SecretVault
from pdv.sync.outbox import OutboxReader

logger = logging.getLogger(__name__)

#: Nome do segredo no cofre. O token nunca entra em `device_settings`.
SYNC_TOKEN_NAME = "sync_token"

#: Códigos são digitados por quem está no balcão, não colados. Aceitamos
#: separadores e caixa livre para que "a1b2-c3d4" e "A1B2C3D4" sejam o mesmo.
_CODE_CLEANUP = re.compile(r"[^A-Za-z0-9]")

MIN_CODE_LENGTH = 6
MAX_CODE_LENGTH = 32


class ActivationError(PdvError):
    """Falha na ativação do terminal."""


class ActivationRefused(ActivationError):
    """O servidor recusou o código (expirado, já usado ou inexistente)."""


class ActivationBlocked(ActivationError):
    """A ativação é possível, mas seria destrutiva agora."""


@dataclass(frozen=True, slots=True)
class ActivationResult:
    tenant_id: str
    store_id: str
    device_id: str
    sync_token: str
    store_name: str = ""
    cloud_base_url: str = ""


class ActivationTransport(Protocol):
    """Contrato do canal de ativação. Injetável para testar sem rede."""

    def activate(self, code: str, fingerprint: dict[str, str]) -> ActivationResult: ...


def normalize_code(raw: str) -> str:
    """Limpa o código digitado.

    Raises:
        ActivationError: se o que sobrou não tem tamanho plausível.
    """
    cleaned = _CODE_CLEANUP.sub("", raw).upper()
    if not MIN_CODE_LENGTH <= len(cleaned) <= MAX_CODE_LENGTH:
        raise ActivationError(
            f"Código de ativação inválido: esperados entre {MIN_CODE_LENGTH} e "
            f"{MAX_CODE_LENGTH} caracteres, recebidos {len(cleaned)}."
        )
    return cleaned


#: O endereço de exemplo do `AppConfig`. Nunca é uma retaguarda de verdade: um
#: terminal instalado que ainda o carrega nunca conseguiria ativar, e por muito
#: tempo foi exatamente isso que o instalador tentava.
PLACEHOLDER_CLOUD_URL = "https://api.erpfood.local"

_LOCAL_HOSTS = frozenset({"localhost", "127.0.0.1", "::1"})


def normalize_server_url(raw: str) -> str:
    """O endereço da retaguarda como o terminal usa: `https://host[/...]/api`.

    O lojista digita o que o painel mostra na barra do navegador —
    `painel.minhaloja.com.br`, com ou sem `https://`, com ou sem barra no fim.
    As rotas do terminal vivem sob `/api`, então o sufixo é acrescentado aqui e
    não exigido de quem digita.

    HTTPS é obrigatório fora desta própria máquina: é por este endereço que o
    token de sincronização e todas as vendas vão trafegar. `http://` só para
    `localhost`, que é o caso de teste com a retaguarda no mesmo computador.

    Raises:
        ActivationError: endereço vazio, malformado ou sem HTTPS.
    """
    from urllib.parse import urlsplit, urlunsplit

    text = raw.strip()
    if not text:
        raise ActivationError("Informe o endereço da retaguarda.")
    if "://" not in text:
        text = "https://" + text

    parts = urlsplit(text)
    host = (parts.hostname or "").lower()
    if parts.scheme not in ("http", "https") or not host or " " in text:
        raise ActivationError(f"Endereço inválido: {raw.strip()!r}.")
    if parts.username or parts.password:
        raise ActivationError("O endereço não pode conter usuário ou senha.")
    if parts.scheme == "http" and host not in _LOCAL_HOSTS:
        raise ActivationError(
            "Use https:// — por este endereço vão passar o token do terminal e "
            "as vendas. http:// só é aceito para localhost."
        )

    path = parts.path.rstrip("/")
    if not path.endswith("/api"):
        path += "/api"
    return urlunsplit((parts.scheme, parts.netloc.lower(), path, "", ""))


def display_server_url(api_url: str) -> str:
    """O endereço como o lojista o reconhece: sem o `/api` do final."""
    if not api_url or api_url == PLACEHOLDER_CLOUD_URL:
        return ""
    return api_url[: -len("/api")] if api_url.endswith("/api") else api_url


def machine_fingerprint() -> dict[str, str]:
    """Identificação da máquina, para o painel mostrar qual terminal é qual.

    É um rótulo de conveniência, **não** um mecanismo de segurança: tudo aqui é
    forjável por quem controla o processo. Quem autentica é o token.
    """
    return {
        "hostname": platform.node(),
        "os": f"{platform.system()} {platform.release()}",
        "arch": platform.machine(),
    }


class HttpActivationTransport:
    """Fala com `POST {cloud}/devices/activate`."""

    def __init__(self, base_url: str, timeout_seconds: float = 20.0) -> None:
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout_seconds

    def activate(self, code: str, fingerprint: dict[str, str]) -> ActivationResult:
        try:
            import httpx
        except ImportError as exc:  # pragma: no cover
            raise ActivationError(
                "httpx não instalado — execute: pip install httpx"
            ) from exc

        try:
            response = httpx.post(
                f"{self._base_url}/devices/activate",
                json={"activation_code": code, "fingerprint": fingerprint},
                timeout=self._timeout,
            )
        except Exception as exc:  # noqa: BLE001 - httpx tem muitas subclasses
            raise ActivationError(
                f"Não foi possível falar com a retaguarda: {exc}"
            ) from exc

        # 4xx é veredito: reenviar o mesmo código dá o mesmo resultado. 5xx é
        # transitório e vale nova tentativa — a distinção evita que o técnico
        # queime um código válido insistindo contra um servidor fora do ar.
        if response.status_code in (400, 401, 403, 404, 409, 410):
            raise ActivationRefused(_refusal_message(response))
        if response.status_code >= 400:
            raise ActivationError(
                f"A retaguarda respondeu {response.status_code}. Tente novamente "
                "em alguns minutos."
            )

        return _parse_activation(response.json(), self._base_url)


def _refusal_message(response: object) -> str:
    detail = ""
    try:
        body = response.json()  # type: ignore[attr-defined]
        if isinstance(body, dict):
            detail = str(body.get("detail") or body.get("message") or "")
    except Exception:  # noqa: BLE001
        detail = ""
    return detail or (
        "Código recusado. Confira se foi digitado corretamente e se ainda não "
        "expirou — gere um novo no painel administrativo."
    )


def _parse_activation(body: object, base_url: str) -> ActivationResult:
    if not isinstance(body, dict):
        raise ActivationError("Resposta de ativação em formato inesperado.")

    required = ("tenant_id", "store_id", "device_id", "sync_token")
    missing = [field for field in required if not body.get(field)]
    if missing:
        raise ActivationError(
            f"Resposta de ativação incompleta: falta {', '.join(missing)}."
        )

    return ActivationResult(
        tenant_id=str(body["tenant_id"]),
        store_id=str(body["store_id"]),
        device_id=str(body["device_id"]),
        sync_token=str(body["sync_token"]),
        store_name=str(body.get("store_name") or ""),
        cloud_base_url=str(body.get("cloud_base_url") or base_url),
    )


# --------------------------------------------------------------------------- #
# Orquestração
# --------------------------------------------------------------------------- #


def activate(
    code: str,
    *,
    database: Database,
    vault: SecretVault,
    transport: ActivationTransport,
    force: bool = False,
) -> ActivationResult:
    """Ativa o terminal e persiste a identidade recebida.

    Args:
        force: permite reapontar um terminal já ativado para outro tenant.
            Só use com a fila vazia e com decisão consciente do suporte.

    Raises:
        ActivationError: código inválido ou falha de comunicação.
        ActivationRefused: o servidor recusou o código.
        ActivationBlocked: a troca destruiria a rastreabilidade de vendas
            ainda não sincronizadas.
    """
    normalized = normalize_code(code)
    store = SettingsStore(database)
    current = store.load()

    result = transport.activate(normalized, machine_fingerprint())

    changing_tenant = (
        current.tenant_id is not None and current.tenant_id != result.tenant_id
    )
    if changing_tenant and not force:
        pending = OutboxReader(database).pending_count()
        if pending:
            raise ActivationBlocked(
                f"Este terminal tem {pending} registro(s) ainda não sincronizado(s) "
                f"com a loja atual. Sincronize antes de reapontá-lo para outra "
                f"loja — caso contrário essas vendas seriam enviadas para o CNPJ "
                f"errado."
            )

    # Token primeiro: gravar as configurações antes deixaria o terminal
    # "ativado" sem credencial se a escrita do cofre falhasse, e ele tentaria
    # sincronizar contra um 401 eterno.
    vault.store(SYNC_TOKEN_NAME, result.sync_token.encode("utf-8"))

    store.set_many(
        {
            "device.tenant_id": result.tenant_id,
            "device.store_id": result.store_id,
            "device.id": result.device_id,
            "device.activated": "1",
            "cloud.base_url": result.cloud_base_url,
            # O nome aparece no login, no título da janela e no cupom. Sem
            # gravá-lo, um terminal ativado seguia dizendo "Confeitaria Demo".
            **({"store.name": result.store_name} if result.store_name else {}),
        }
    )

    logger.info(
        "Terminal ativado: device=%s tenant=%s loja=%r",
        result.device_id,
        result.tenant_id,
        result.store_name,
    )
    return result


def is_activated(database: Database) -> bool:
    return SettingsStore(database).load().activated


def load_sync_token(vault: SecretVault) -> str | None:
    """Token de sincronização em texto, ou `None` se o terminal não foi ativado."""
    raw = vault.load(SYNC_TOKEN_NAME)
    return raw.decode("utf-8") if raw is not None else None

File: apps/pdv-net/src/Pdv.Data/Provisioning/Activation.cs
using System.Net.Http.Json;
using System.Runtime.InteropServices;
using System.Text.Json;
using System.Text.Json.Serialization;
using Pdv.Core;
using Pdv.Data.Secrets;

namespace Pdv.Data.Provisioning;

/// <summary>Código inválido ou falha de comunicação: a mensagem vai para o balcão.</summary>
public class ActivationException(string message, Exception? inner = null) : Exception(message, inner);

/// <summary>A retaguarda recusou o código (4xx): reenviar o mesmo dá o mesmo resultado.</summary>
public sealed class ActivationRefusedException(string message) : ActivationException(message);

/// <summary>A troca de loja levaria vendas ainda não sincronizadas para o CNPJ errado.</summary>
public sealed class ActivationBlockedException(string message) : ActivationException(message);

/// <summary>O que a retaguarda devolve ao trocar o código pelo token.</summary>
public sealed record ActivationResult(
    string TenantId,
    string StoreId,
    string DeviceId,
    string SyncToken,
    string StoreName,
    string CloudBaseUrl);

/// <summary>Rótulo da máquina para o painel. Conveniência, não segurança: quem autentica é o token.</summary>
public sealed record MachineFingerprint(
    [property: JsonPropertyName("hostname")] string Hostname,
    [property: JsonPropertyName("os")] string Os,
    [property: JsonPropertyName("arch")] string Arch)
{
    /// <summary>Os mesmos rótulos do <c>platform</c> do Python: <c>Windows 11</c>, <c>AMD64</c>.</summary>
    public static MachineFingerprint Current() => new(
        Environment.MachineName,
        $"Windows {(Environment.OSVersion.Version.Build >= 22000 ? "11" : Environment.OSVersion.Version.Major.ToString())}",
        RuntimeInformation.OSArchitecture switch
        {
            Architecture.X64 => "AMD64",
            Architecture.Arm64 => "ARM64",
            Architecture.X86 => "x86",
            var other => other.ToString(),
        });
}

public interface IActivationTransport
{
    Task<ActivationResult> ActivateAsync(
        string code, MachineFingerprint fingerprint, byte[] deviceSecret, CancellationToken cancellation);
}

/// <summary>
/// Endereço e código como o <c>pdv.provisioning.activation</c> do Python os trata,
/// regra por regra (<c>contracts/activation.json</c>).
/// </summary>
public static class Activation
{
    /// <summary>O nome do token de sincronização no cofre.</summary>
    public const string SyncTokenName = "sync_token";

    public const int MinCodeLength = 6;
    public const int MaxCodeLength = 32;

    /// <summary>O endereço de exemplo do <c>AppConfig</c>; nunca é uma retaguarda de verdade.</summary>
    public const string PlaceholderCloudUrl = "https://api.erpfood.local";

    private static readonly HashSet<string> LocalHosts = new(StringComparer.Ordinal) { "localhost", "127.0.0.1", "::1" };

    /// <summary>
    /// Só letras e números ASCII, em maiúsculas — o painel mostra <c>ABCD-EFGH-JKMN</c>
    /// e o técnico digita como quiser.
    /// </summary>
    /// <exception cref="ActivationException">Sobrou um tamanho implausível.</exception>
    public static string NormalizeCode(string raw)
    {
        var cleaned = new string(raw.Where(char.IsAsciiLetterOrDigit).ToArray()).ToUpperInvariant();
        if (cleaned.Length is < MinCodeLength or > MaxCodeLength)
        {
            throw new ActivationException(
                $"Código de ativação inválido: esperados entre {MinCodeLength} e {MaxCodeLength} " +
                $"caracteres, recebidos {cleaned.Length}.");
        }
        return cleaned;
    }

    /// <summary>
    /// O endereço da retaguarda como o terminal usa: <c>https://host[/...]/api</c>.
    /// </summary>
    /// <remarks>
    /// O lojista digita o que vê no navegador, com ou sem <c>https://</c> e barra
    /// no fim. HTTPS é obrigatório fora desta máquina: por aqui passam o token e as
    /// vendas. Analisado à mão, e não com <see cref="Uri"/>, porque o <c>Uri</c>
    /// reescreve o endereço (tira a porta 443, converte IDN) e o Python não.
    /// </remarks>
    /// <exception cref="ActivationException">Vazio, malformado ou sem HTTPS.</exception>
    public static string NormalizeServerUrl(string raw)
    {
        var text = raw.Trim();
        if (text.Length == 0) throw new ActivationException("Informe o endereço da retaguarda.");
        if (!text.Contains("://", StringComparison.Ordinal)) text = "https://" + text;

        var separator = text.IndexOf("://", StringComparison.Ordinal);
        var scheme = text[..separator].ToLowerInvariant();
        var rest = text[(separator + 3)..];
        var netlocEnd = rest.IndexOfAny(['/', '?', '#']);
        var netloc = netlocEnd < 0 ? rest : rest[..netlocEnd];
        var tail = netlocEnd < 0 ? "" : rest[netlocEnd..];
        var pathEnd = tail.IndexOfAny(['?', '#']);
        var path = pathEnd < 0 ? tail : tail[..pathEnd];

        var hasUserInfo = netloc.Contains('@');
        var host = HostOf(hasUserInfo ? netloc[(netloc.LastIndexOf('@') + 1)..] : netloc);

        if (scheme is not ("http" or "https") || host.Length == 0 || text.Contains(' '))
        {
            throw new ActivationException($"Endereço inválido: '{raw.Trim()}'.");
        }
        if (hasUserInfo) throw new ActivationException("O endereço não pode conter usuário ou senha.");
        if (scheme == "http" && !LocalHosts.Contains(host))
        {
            throw new ActivationException(
                "Use https:// — por este endereço vão passar o token do terminal e as vendas. " +
                "http:// só é aceito para localhost.");
        }

        path = path.TrimEnd('/');
        if (!path.EndsWith("/api", StringComparison.Ordinal)) path += "/api";
        return $"{scheme}://{netloc.ToLowerInvariant()}{path}";
    }

    private static string HostOf(string hostAndPort)
    {
        if (hostAndPort.StartsWith('['))
        {
            var close = hostAndPort.IndexOf(']');
            return (close < 0 ? hostAndPort[1..] : hostAndPort[1..close]).ToLowerInvariant();
        }
        var colon = hostAndPort.IndexOf(':');
        return (colon < 0 ? hostAndPort : hostAndPort[..colon]).ToLowerInvariant();
    }

    /// <summary>O endereço como o lojista o reconhece: sem o <c>/api</c> do final.</summary>
    public static string DisplayServerUrl(string? apiUrl)
    {
        if (string.IsNullOrEmpty(apiUrl) || apiUrl == PlaceholderCloudUrl) return "";
        return apiUrl.EndsWith("/api", StringComparison.Ordinal) ? apiUrl[..^"/api".Length] : apiUrl;
    }

    /// <summary>A raiz <c>/api</c> da nuvem, aceite quem digitou a origem ou a raiz.</summary>
    public static string CloudApiRoot(string baseUrl)
    {
        var trimmed = baseUrl.TrimEnd('/');
        return trimmed.EndsWith("/api", StringComparison.Ordinal) ? trimmed : trimmed + "/api";
    }

    /// <summary>
    /// Ativa o terminal e grava a identidade recebida — no banco e no cofre que
    /// o PDV em Python também lê.
    /// </summary>
    /// <param name="force">Reapontar para outro tenant mesmo com a fila cheia. Só o suporte.</param>
    /// <exception cref="ActivationException">Código inválido ou falha de comunicação.</exception>
    /// <exception cref="ActivationRefusedException">A retaguarda recusou o código.</exception>
    /// <exception cref="ActivationBlockedException">A troca levaria vendas para o CNPJ errado.</exception>
    public static async Task<ActivationResult> ActivateAsync(
        string code,
        PdvDatabase database,
        SecretVault vault,
        IActivationTransport transport,
        bool force = false,
        TimeProvider? clock = null,
        CancellationToken cancellation = default)
    {
        var normalized = NormalizeCode(code);
        var currentTenant = database.Scalar(
            "SELECT value FROM device_settings WHERE key = 'device.tenant_id'") as string;

        var result = await transport.ActivateAsync(
            normalized, MachineFingerprint.Current(), vault.EnsureDeviceSecret(), cancellation);

        var changingTenant = !string.IsNullOrEmpty(currentTenant) && currentTenant != result.TenantId;
        if (changingTenant && !force)
        {
            var pending = Convert.ToInt64(database.Scalar("SELECT COUNT(*) FROM sync_outbox"));
            if (pending > 0)
            {
                throw new ActivationBlockedException(
                    $"Este terminal tem {pending} registro(s) ainda não sincronizado(s) com a loja atual. " +
                    "Sincronize antes de reapontá-lo para outra loja — caso contrário essas vendas seriam " +
                    "enviadas para o CNPJ errado.");
            }
        }

        // Token primeiro: gravar as configurações antes deixaria o terminal
        // "ativado" sem credencial se o cofre falhasse, sincronizando contra um
        // 401 eterno.
        vault.Store(SyncTokenName, System.Text.Encoding.UTF8.GetBytes(result.SyncToken));

        var values = new List<(string Key, string Value)>
        {
            ("device.tenant_id", result.TenantId),
            ("device.store_id", result.StoreId),
            ("device.id", result.DeviceId),
            ("device.activated", "1"),
            ("cloud.base_url", result.CloudBaseUrl),
        };
        // O nome aparece no login e no cupom; sem ele o terminal seguia "Confeitaria Demo".
        if (result.StoreName.Length > 0) values.Add(("store.name", result.StoreName));

        var now = Iso.Now(clock);
        database.InTransaction(transaction =>
        {
            foreach (var (key, value) in values)
            {
                using var command = transaction.Command(
                    "INSERT INTO device_settings (key, value, updated_at) VALUES ($key, $value, $now) " +
                    "ON CONFLICT (key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at",
                    ("$key", key), ("$value", value), ("$now", now));
                command.ExecuteNonQuery();
            }
        });
        return result;
    }
}

/// <summary>Fala com <c>POST {nuvem}/api/devices/activate</c>.</summary>
public sealed class HttpActivationTransport(string baseUrl, HttpClient? client = null) : IActivationTransport
{
    private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(20);

    private readonly HttpClient _client = client ?? new HttpClient { Timeout = Timeout };
    private readonly string _baseUrl = baseUrl.TrimEnd('/');

    public async Task<ActivationResult> ActivateAsync(
        string code, MachineFingerprint fingerprint, byte[] deviceSecret, CancellationToken cancellation)
    {
        int status;
        string body;
        try
        {
            using var response = await _client.PostAsJsonAsync(
                $"{Activation.CloudApiRoot(_baseUrl)}/devices/activate",
                new ActivationRequest(code, fingerprint, Convert.ToHexStringLower(deviceSecret)),
                cancellation);
            status = (int)response.StatusCode;
            body = await response.Content.ReadAsStringAsync(cancellation);
        }
        catch (Exception error) when (error is HttpRequestException or TaskCanceledException)
        {
            // A mensagem de fora do HttpClient é genérica ("Error while copying
            // content to a stream"); a causa que o suporte precisa está dentro.
            var cause = error;
            while (cause.InnerException is not null) cause = cause.InnerException;
            var detail = ReferenceEquals(cause, error) ? error.Message : $"{error.Message} ({cause.Message})";
            throw new ActivationException($"Não foi possível falar com a retaguarda: {detail}", error);
        }

        // 4xx é veredito; 5xx é transitório. A distinção evita que o técnico
        // queime um código válido insistindo contra um servidor fora do ar.
        if (status is 400 or 401 or 403 or 404 or 409 or 410)
        {
            throw new ActivationRefusedException(RefusalMessage(body));
        }
        if (status >= 400)
        {
            throw new ActivationException(
                $"A retaguarda respondeu {status}. Tente novamente em alguns minutos.");
        }
        return Parse(body, _baseUrl);
    }

    private static string RefusalMessage(string body)
    {
        try
        {
            using var document = JsonDocument.Parse(body);
            if (document.RootElement.ValueKind == JsonValueKind.Object)
            {
                foreach (var field in new[] { "detail", "message" })
                {
                    if (document.RootElement.TryGetProperty(field, out var value) &&
                        value.ValueKind == JsonValueKind.String && value.GetString() is { Length: > 0 } text)
                    {
                        return text;
                    }
                }
            }
        }
        catch (JsonException)
        {
        }
        return "Código recusado. Confira se foi digitado corretamente e se ainda não expirou — " +
               "gere um novo no painel administrativo.";
    }

    internal static ActivationResult Parse(string body, string baseUrl)
    {
        JsonElement root;
        try
        {
            using var document = JsonDocument.Parse(body);
            root = document.RootElement.Clone();
        }
        catch (JsonException)
        {
            throw new ActivationException("Resposta de ativação em formato inesperado.");
        }
        if (root.ValueKind != JsonValueKind.Object)
        {
            throw new ActivationException("Resposta de ativação em formato inesperado.");
        }

        string Text(string name) =>
            root.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
                ? value.GetString() ?? ""
                : "";

        var missing = new[] { "tenant_id", "store_id", "device_id", "sync_token" }
            .Where(name => Text(name).Length == 0).ToArray();
        if (missing.Length > 0)
        {
            throw new ActivationException($"Resposta de ativação incompleta: falta {string.Join(", ", missing)}.");
        }
        var cloud = Text("cloud_base_url");
        return new ActivationResult(
            Text("tenant_id"), Text("store_id"), Text("device_id"), Text("sync_token"),
            Text("store_name"), cloud.Length > 0 ? cloud : baseUrl);
    }

    private sealed record ActivationRequest(
        [property: JsonPropertyName("activation_code")] string ActivationCode,
        [property: JsonPropertyName("fingerprint")] MachineFingerprint Fingerprint,
        [property: JsonPropertyName("device_secret_hex")] string DeviceSecretHex);
}
